Automattic/wp-calypso
# E2E: Add base image for running core GB tests on dotcom

This is a follow-up to:
* https://github.com/Automattic/wp-calypso/pull/86949/
* pMz3w-jlo-p2

## Why

**Quick recap:**

This is part of a greater effort (to try) to run GB core tests in dotcom. Not all tests will be feasible, but we think that we can run a selection of them and reduce the reliance on tests implemented in dotcom (Calypso). It would also improve the testing coverage, as core is continuously adding/mantaining tests as they work on Gutenberg, and the changes in E2Es in Calypso (at least the ones related to GB) are very reactive.

@WunderBart worked on a PoC that I merged a couple of weeks ago: https://github.com/Automattic/wp-calypso/pull/86949/. This added a new build in `Calypso->WPCom Tests->Gutenberg Core E2E Tests`. The build started to fail because of node/npm version mismatches, though, and I created a systemsrequest to gather insights about it: pMz3w-jlo-p2.

I clarified in that request that I needed to create a new base image for this build. This is what I'm doing in this PR. This image already has the right version of node/npm for Gutenberg, and the test task already runs when testing locally, though tests fail becaus the secrets env variables are not available (these: https://github.com/Automattic/wp-calypso/pull/86949/files#diff-f142e837d6e2259b594adbf3137e36f03c8c122468a4f17e4fc07d858d660b93R91-R99).

**So, the purpose of this changeset is to solve the node/npm version mismatch.**

## How?

By creating a new image that inherits from a base node-20.11.11 image. It has a similar setup to the ci-e2e image, but doesn't inherit from it. INheriting from ci-e2e didn't make much sense: It's made for Playwright, yes, but Playwright in Calypso. Also, it inherited (indirectly) from an older node base image, and I didn't want to change the base image as it would affect all the images defined in `Dockerfile.base`. 

## Next steps

Now that the image has been proven to work, the next steps are:
* [x] Push this to the A8C registry. I never did that, and I will probably submit a systemsrequest just in case, to check if there are anything I should be aware of. I already authenticated to the A8C registry, though, so I should be ready to push it (this can be done by running the `test/build-tag-push-ci-e2e-gb-core-on-dotcom-base-image.sh` script);
* [x] After it's in the registry, modify the build definition here to use it: https://github.com/Automattic/wp-calypso/pull/86949/files#diff-f142e837d6e2259b594adbf3137e36f03c8c122468a4f17e4fc07d858d660b93R74. An example of how to set the base docker image used for the TC agent: https://href.li/?https://github.com/Automattic/wp-calypso/blob/trunk/.teamcity/_self/lib/customBuildType/E2EBuildType.kt#L162. This will be done in another follow-up PR.

## Testing Instructions

1. Checkout this branch;
2. Make sure you have docker installed and properly setup;
3. From the Calypso repo root, run './run-gb-core-tests-on-dotcom.sh'. This script emulates what the build would run here if the build were running from TC: https://github.com/Automattic/wp-calypso/pull/86949/files#diff-f142e837d6e2259b594adbf3137e36f03c8c122468a4f17e4fc07d858d660b93R102-R140
4. It should properly create the images and will eventually fail with an output similar to the following:

```
316.5 Webkit 17.4 (playwright build v1921) downloaded to /root/.cache/ms-playwright/webkit-1921
316.9 Error: WP_CLIENT_ID and WP_CLIENT_SECRET environment variables must be set.
316.9
316.9    at ../../../packages/e2e-test-utils-playwright/src/request-utils/login.ts:16
316.9
316.9   14 |    const clientSecret = process.env.WP_CLIENT_SECRET;
316.9   15 |    if ( ! clientId || ! clientSecret ) {
316.9 > 16 |            throw new Error(
316.9      |                  ^
316.9   17 |                    'WP_CLIENT_ID and WP_CLIENT_SECRET environment variables must be set.'
316.9   18 |            );
316.9   19 |    }
316.9
316.9     at RequestUtils.login (/workspace/gutenberg/packages/e2e-test-utils-playwright/src/request-utils/login.ts:16:9)
316.9     at RequestUtils.setupRest (/workspace/gutenberg/packages/e2e-test-utils-playwright/src/request-utils/rest.ts:25:13)
316.9     at globalSetup (/workspace/gutenberg/test/e2e/config/global-setup.ts:26:21)
316.9
------
ci-e2e-gb-core-on-dotcom-test.dockerfile:19
--------------------
  18 |     # Clone Gutenberg and prepare it for the E2E tests, and run them
  19 | >>> RUN git clone --depth 1 -b try/run-e2e-tests-against-wpcom https://github.com/WordPress/gutenberg.git /workspace/gutenberg && \
  20 | >>>     cd /workspace/gutenberg && \
  21 | >>>     npm ci && \
  22 | >>>     npm run build:packages && \
  23 | >>>              npm run test:e2e:playwright
  24 |
--------------------
``` 

That's expected, and means the goals of this changeset have been reached. It's failing because it needs some secrets that are not set here. Once the right build uses this image, the env vars will be set, and it should work fine.
